flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnalignerTest.java

```diff
@@ -512,6 +512,123 @@ public void testConcurrentProcessBarrierAndNotifyBarrierReceived() throws Except
+
+		assertFalse(handler.isCheckpointPending());
+		assertTrue(unaligner.isCheckpointPending());
+		assertEquals(-1L, handler.getLatestCheckpointId());
+		assertEquals(checkpointId, unaligner.getCurrentCheckpointId());
+
+		testProcessCancellationBarrier(handler, invokable, checkpointId);
+	}
+
+	/**
+	 * Tests {@link CheckpointBarrierUnaligner#processCancellationBarrier(CancelCheckpointMarker)}
+	 * abort the current pending checkpoint triggered by
+	 * {@link CheckpointBarrierUnaligner#processBarrier(CheckpointBarrier, int)}.
+	 */
+	@Test
+	public void testProcessCancellationBarrierWitchPendingCheckpoint2() throws Exception {
+		final long checkpointId = 0L;
+		final ValidatingCheckpointInvokable invokable = new ValidatingCheckpointInvokable();
+		final CheckpointBarrierUnaligner handler = new CheckpointBarrierUnaligner(
+			new int[] { 1 }, ChannelStateWriter.NO_OP, "test", invokable);
+
+		// should trigger respective checkpoint
+		handler.processBarrier(buildCheckpointBarrier(checkpointId), 0);
+
+		assertTrue(handler.isCheckpointPending());
+		assertTrue(handler.getThreadSafeUnaligner().isCheckpointPending());
+		assertEquals(checkpointId, handler.getLatestCheckpointId());
+		assertEquals(checkpointId, handler.getThreadSafeUnaligner().getCurrentCheckpointId());
+
+		testProcessCancellationBarrier(handler, invokable, checkpointId);
+	}
+
+	@Test
+	public void testProcessCancellationBarrierBeforeProcessAndReceiveBarrier() throws Exception {
+		final long checkpointId = 0L;
+		final ValidatingCheckpointInvokable invokable = new ValidatingCheckpointInvokable();
+		final CheckpointBarrierUnaligner handler = new CheckpointBarrierUnaligner(
+			new int[] { 1 }, ChannelStateWriter.NO_OP, "test", invokable);
+
+		handler.processCancellationBarrier(new CancelCheckpointMarker(checkpointId));
+
+		verifyTriggeredCheckpoint(handler, invokable, checkpointId);
+
+		// it would not trigger checkpoint since the respective cancellation barrier already happened before
+		handler.processBarrier(buildCheckpointBarrier(checkpointId), 0);
+		handler.getThreadSafeUnaligner().notifyBarrierReceived(buildCheckpointBarrier(checkpointId), new InputChannelInfo(0, 0));
+
+		verifyTriggeredCheckpoint(handler, invokable, checkpointId);
+	}
+
+	private void testProcessCancellationBarrier(
+			CheckpointBarrierUnaligner handler,
+			ValidatingCheckpointInvokable invokable,
+			long currentCheckpointId) throws Exception {
+
+		// should abort current checkpoint while processing CancelCheckpointMarker
+		handler.processCancellationBarrier(new CancelCheckpointMarker(currentCheckpointId));
+		verifyTriggeredCheckpoint(handler, invokable, currentCheckpointId);
+
+		final long canceledCheckpointId = 1L;
+		// should update current checkpoint id and abort notification while processing CancelCheckpointMarker
+		handler.processCancellationBarrier(new CancelCheckpointMarker(canceledCheckpointId));
+		verifyTriggeredCheckpoint(handler, invokable, canceledCheckpointId);
```

Comment:
Would it make sense to extract these assertions in separate test cases? There might be a bias in the implementation, when the checkpoint has already been canceled. It would only add two test cases afaik.

Reply:
Makes sense.

Reply:
I take a random way to cover both cases for avoiding some duplicated codes.